File: app/src/main/java/com/ardev/idroid/ui/main/MainActivity.java
package com.ardev.idroid.ui.main;


import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ObjectAnimator;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.TextUtils;
import com.ardev.component.widget.breadcrumbs.model.IBreadcrumbItem;
import com.ardev.idroid.ext.JsonUtils;
import com.ardev.idroid.service.IndexServiceConnection;
import com.ardev.idroid.service.IndexService;
import com.ardev.idroid.ui.home.ProjectProvider;
import androidx.annotation.NonNull;
import com.ardev.builder.project.Project;
import com.ardev.idroid.ext.ProjectUtils;
import com.ardev.idroid.ui.home.ProjectManager;
import com.google.android.material.transition.MaterialFade;
import android.view.ViewGroup;
import androidx.transition.TransitionManager;
import com.ardev.idroid.ui.main.model.Editable;
import com.ardev.idroid.ui.main.adapter.PageAdapter;
import com.ardev.idroid.databinding.ActivityMainBinding;
import com.ardev.idroid.ui.main.adapter.FileListAdapter;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import com.google.gson.reflect.TypeToken;
import com.ardev.idroid.ext.UtilKt;
import com.ardev.idroid.app.MainViewModelProvider;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Toast;
import androidx.appcompat.widget.PopupMenu;
import android.view.animation.AnticipateInterpolator;
import android.widget.Toast;
import androidx.annotation.Nullable;
import androidx.core.graphics.Insets;
import androidx.core.view.GravityCompat;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.lifecycle.Observer;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.lifecycle.SavedStateViewModelFactory;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager2.widget.ViewPager2;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.ardev.component.widget.breadcrumbs.BreadcrumbsView;
import com.google.android.material.tabs.TabLayoutMediator;
import java.lang.reflect.Type;
import com.ardev.idroid.base.BaseActivity;
import com.google.gson.Gson;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;
import org.apache.commons.io.FileUtils;

import java.util.concurrent.Executors;
import com.google.android.material.appbar.MaterialToolbar;
import java.io.File;
import androidx.viewpager2.widget.CompositePageTransformer;
import com.ardev.component.widget.breadcrumbs.DefaultBreadcrumbsCallback;
import com.google.android.material.tabs.TabLayout;
import com.ardev.f.storage.Storage;
import com.ardev.component.widget.breadcrumbs.model.BreadcrumbItem;
import com.ardev.idroid.R;

public class MainActivity extends  BaseActivity implements ProjectManager.OnProjectOpenListener {
	
	
FileListAdapter mFilesAdapter;
PageAdapter mPagerAdapter;
MainViewModel mainViewModel;
ActivityMainBinding v;
Project project;
private ProjectManager mProjectManager;
private IndexServiceConnection mIndexServiceConnection;
Storage st;

boolean isFirst = true, isFirst2 = true, isFirst3 = true;


	@Override
	protected void onCreate(Bundle bundle) {
		super.onCreate(bundle);
v = ActivityMainBinding.inflate(getLayoutInflater());
		setContentView(v.getRoot());
	 st = new Storage(this);	
	 mainViewModel = new ViewModelProvider(this, new SavedStateViewModelFactory(getApplication(), this)).get(MainViewModel.class);
	MainViewModelProvider.init(mainViewModel);
	 project = new Project(ProjectProvider.getRootFile());
	 
	 mProjectManager = ProjectManager.getInstance();
     mProjectManager.addOnProjectOpenListener(this);
	  mIndexServiceConnection = new IndexServiceConnection(mainViewModel);
    
	
		init();
		initBCV();
		initRV();
		initTab();
		initObserver();
		mainViewModel.setFatherPath(project.getRootFile().getPath());
		mainViewModel.init();
		
	if (!project.equals(mProjectManager.getCurrentProject())) openProject(project);
	}
	
	private void init() {
UtilKt.addSystemWindowInsetToPadding(v.navView, false, false, false, false);
UtilKt.addSystemWindowInsetToMargin(v.pager2, false, false, false, false);

ViewCompat.setOnApplyWindowInsetsListener(
                v.pager2,
                (vi, insets) -> {
                    boolean imeVisible = insets.isVisible(WindowInsetsCompat.Type.ime());

                    Insets in = insets.getInsets(WindowInsetsCompat.Type.systemBars());
                   v.pager2.setPadding(0, 0, 0, in.bottom);
                    if (imeVisible) {
                        int imeHeight = insets.getInsets(WindowInsetsCompat.Type.ime()).bottom;
                        v.pager2.setPadding(0, 0, 0, imeHeight);
                    }
                    return insets;
        });



	setSupportActionBar(v.toolbar);
	getSupportActionBar().setDisplayHomeAsUpEnabled(true);
		
		ActionBarDrawerToggle toggle =
		new ActionBarDrawerToggle(
		this,
		v.drawer,
		v.toolbar,
		R.string.app_name,
		R.string.app_name);
	
		toggle.syncState();
		toggle.setDrawerSlideAnimationEnabled(true);
		v.drawer.addDrawerListener(toggle);
		}
		

	private void initBCV() {
	
 	v.breadcrumbsView.setCallback( new DefaultBreadcrumbsCallback<BreadcrumbItem>() {
	@Override
	public void onNavigateBack(BreadcrumbItem item, int position) {
		
		mainViewModel.setCurrentPath(getPath(position));
		 bkpListBreadcrumb(v.breadcrumbsView.getItems());
	}
	
	@Override
	public void onNavigateNewLocation(BreadcrumbItem newItem, int changedPosition) {
		
	}
});
		
		}
		
		private void initRV(){

mFilesAdapter = new FileListAdapter();
  v.list.setAdapter(mFilesAdapter);
v.list.setHasFixedSize(true);
v.list.setItemAnimator(new DefaultItemAnimator());
mFilesAdapter.setOnClickListener( (view, file) -> {
			 if (file.isDirectory()) {
  	mainViewModel.addBreadcrumbItem(file.getPath());
	mainViewModel.setCurrentPath(file.getPath());
		} else {
	mainViewModel.addEditor(new Editable(file, false));
v.drawer.closeDrawer(GravityCompat.START);

}
});

mFilesAdapter.setOnLongClickListener( (v, file) -> { 

if (file.isDirectory()) showOptionsMenu(v, file);
else if((ProjectUtils.isLayoutXMLFile(file) || ProjectUtils.isValuesXMLFile(file)  || ProjectUtils.isDrawableXMLFile(file)) && file.isFile()) showFilesMenu(v, file);

});
 
 
 
 }
 
 	public void initObserver() {
 	
 	mainViewModel.getFatherPath().observe(MainActivity.this, fatherPath -> {
       	getSupportActionBar().setTitle(new File(fatherPath).getName());
       		
       		
      File breadcrumbs = new File(fatherPath + "/app/build/.cache", ".breadcrumbs");
      
if (breadcrumbs.exists()) {
	try {
   List<BreadcrumbItem> listBci =  JsonUtils.jsonToList(FileUtils.readFileToString(breadcrumbs, Charset.defaultCharset()), BreadcrumbItem.class);
   listBci.forEach( it -> v.breadcrumbsView.addItem(it));
  
   	
		} catch (Exception e) {}
		
		
		} else {
	
	v.breadcrumbsView.addItem(BreadcrumbItem.createSimpleItem(new File(fatherPath).getName()));
	
	}

	

            });
	mainViewModel.getCurrentPath().observe(MainActivity.this, path -> { 
	String _path = null;
	if(isFirst3) {
                isFirst3 = false;
     File currentPathCache = new File(mainViewModel.getFatherPath().getValue() + "/app/build/.cache", ".currentpath");
if (currentPathCache.exists()) {
	try {
   _path =  new Gson().fromJson(FileUtils.readFileToString(currentPathCache, Charset.defaultCharset()), String.class);
   
	} catch (Exception e) {} 
	} else {
	_path = path;
	}
	} else {
	_path = path;
	}
	showList(_path);
	bkpCurrentPath(_path);
	});
	
	
	
	
	mainViewModel.getBreadcrumbItem().observe(MainActivity.this, file -> {
		BreadcrumbItem breadcrumbItem = new BreadcrumbItem(mainViewModel.getListDir().getValue());
		breadcrumbItem.setSelectedItem(file.getName());
  v.breadcrumbsView.addItem(breadcrumbItem);
  bkpListBreadcrumb(v.breadcrumbsView.getItems());
	});
		
	mainViewModel.getListEditor()
                .observe(MainActivity.this, files -> {
                
                    mPagerAdapter.submitList(files);
                    v.tab.setVisibility(files.isEmpty() ? View.GONE : View.VISIBLE);
                });
                
   mainViewModel.getCurrentTabPosition()
                .observe(MainActivity.this, position -> {
                if(isFirst) {
                isFirst = false;
                 File positionFile = new File(mainViewModel.getFatherPath().getValue() + "/app/build/.cache", ".position");
if (positionFile.exists()) {
	try {
   int pos=  new Gson().fromJson(FileUtils.readFileToString(positionFile, Charset.defaultCharset()), int.class);
  v.pager2.setCurrentItem(pos, false);
		} catch (Exception e) {} 
		} else {
			 v.pager2.setCurrentItem(position, false);
			}
              } else {
                
             v.pager2.setCurrentItem(position, false);
				
               }
                 });
                 
    mainViewModel.isIndexing()
                .observe(MainActivity.this, indexing -> {
                   
                });
                
        mainViewModel.getCurrentState()
                .observe(MainActivity.this, message -> getSupportActionBar().setSubtitle(message));
		}
		
private String getPath(int depth) {
	if (depth == -1) depth = v.breadcrumbsView.getItems().size() - 1;
	StringBuffer sbPath = new StringBuffer(mainViewModel.getFatherPath().getValue());
	for (int i = 1; i <= depth; i++) {
		
		sbPath.append("/").append(v.breadcrumbsView.getItems().get(i).getSelectedItem());
	}
	return sbPath.toString();
}
 
 private void showList(String path) {
   
    Executors.newSingleThreadExecutor().execute(() -> {
    
        List<File> list = st.getFiles(path);
         
          
       runOnUiThread(() -> {
     
	
                    
            if (list.isEmpty()) {
                v.scrollingView.setVisibility(View.GONE);
                v.emptyItem.setVisibility(View.VISIBLE);
            } else {
                v.scrollingView.setVisibility(View.VISIBLE);
                v.emptyItem.setVisibility(View.GONE);
                 mFilesAdapter.setList(list);
        mainViewModel.setListCurrent(list);
        }
        
            
	
	});
	});
	}

private void showOptionsMenu(View view, File file) {
   
    
    CharSequence[] chars = new CharSequence[]{"Compartir", "Borrar"};
		
		MaterialAlertDialogBuilder b = new MaterialAlertDialogBuilder(MainActivity.this, R.style.ThemeOverlay_MaterialAlertDialog_NoRounded);
		 
		b.setItems(chars, (d, i) -> {

			 
		 });
		 b.show();
        
    }
private void showFilesMenu(View view, File file) {
   
    
    CharSequence[] chars = new CharSequence[]{"Previsualizar", "Editar"};
		
		MaterialAlertDialogBuilder b = new MaterialAlertDialogBuilder(MainActivity.this, R.style.ThemeOverlay_MaterialAlertDialog_NoRounded);
		 
		b.setItems(chars, (d, item) -> {
		switch(item) {
		case 0: 
		mainViewModel.addEditor(new Editable(file, true));
		v.drawer.closeDrawer(GravityCompat.START);
		break;
		case 1:
		mainViewModel.addEditor(new Editable(file, false));
		v.drawer.closeDrawer(GravityCompat.START);
		break;
		}
			 
		 });
		 b.show();
        
    }
    
   private void showTabMenu(TabLayout.Tab tab) {
   Editable editable = mainViewModel.getListEditor().getValue().get(tab.getPosition());
   boolean isPreview = mainViewModel.getListEditor().getValue().get(tab.getPosition()).isPreview();
    File file = mainViewModel.getListEditor().getValue().get(tab.getPosition()).getFile();
    
    String prev =  mainViewModel.getListEditor().getValue().get(tab.getPosition()).getFile().getName();

   PopupMenu menu = new PopupMenu(MainActivity.this, tab.view);
	
		menu.getMenu().add("Cerrar");
		menu.getMenu().add("Cerrar otros");
		menu.getMenu().add("Cerrar todos");
		menu.setOnMenuItemClickListener( item -> {
					 switch(item.getTitle().toString()) {
						 
						 case "Cerrar":
					mainViewModel.closeEditor(editable);
							 break;
							 
							 case "Cerrar otros":
					mainViewModel.closeOthersEditor(editable);
							 break;
							 
						 case "Cerrar todos":
					mainViewModel.closeAllEditor();
							 break;	 
					 }
					
					
					
					return true;
				});
				menu.show();
   
   
   }
    
    
    
    private void initTab() {
    
    
		
   mPagerAdapter = new PageAdapter(getSupportFragmentManager(), getLifecycle());
        v.pager2.setAdapter(mPagerAdapter);
        v.pager2.setUserInputEnabled(false);
		v.pager2.setPageTransformer((view, position) -> {
		view.setAlpha(0.2f);
		view.setVisibility(View.VISIBLE);
		view.animate().alpha(1.0f).setDuration(view.getResources().getInteger(android.R.integer.config_shortAnimTime));
		});
        
        v.tab.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
                
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {
           showTabMenu(tab);
               
            }

            @Override
            public void onTabSelected(TabLayout.Tab tab) {
               mainViewModel.setCurrentTabPosition(tab.getPosition());

            }
        });
		
        new TabLayoutMediator(v.tab, v.pager2, true, false, (tab, i) -> {
			boolean isPreview = mainViewModel.getListEditor().getValue().get(i).isPreview();

			String fileName = mainViewModel.getListEditor().getValue().get(i).getFile().getName();
			
			tab.setText(fileName);
        	if(isPreview) tab.setIcon(R.drawable.ic_preview);
        	 
        	
        }).attach();
    
    
    
    }
    
    
     public void openProject(@NonNull Project project) {
        
        mIndexServiceConnection.setProject(project);

        mainViewModel.setIndexing(true);

        Intent intent = new Intent(MainActivity.this, IndexService.class);
        startService(intent);
        bindService(intent, mIndexServiceConnection, Context.BIND_IMPORTANT);
    }
    
     @Override
    public void onProjectOpen(Project project) {
    
    
    }
    
    
   @Override
        public void onBackPressed() {
             if (v.drawer.isDrawerOpen(GravityCompat.START)) {
            v.drawer.closeDrawer(GravityCompat.START);
            } else {
            super.onBackPressed();
            }
	}
	
	
	public MaterialToolbar getToolbar() {
        return v.toolbar;
    }
	
	  private void bkpListBreadcrumb(List<IBreadcrumbItem> bList) {
   try {
         File bciFile = new File(mainViewModel.getFatherPath().getValue() + "/app/build/.cache", ".breadcrumbs");
         String content  = new GsonBuilder()
                .setPrettyPrinting()
                .create()
                .toJson( bList,new TypeToken<List<BreadcrumbItem>>(){}.getType());
        
        if(!bciFile.exists()) bciFile.createNewFile();
		if(!bList.isEmpty()){
        FileUtils.writeStringToFile(bciFile,
                 content,
                 Charset.defaultCharset());
                 }
                 } catch (Exception e) {}
   
   }
   
   private void bkpCurrentPath(String path) {
   try {
         File pathFile= new File(mainViewModel.getFatherPath().getValue() + "/app/build/.cache", ".currentpath");
        
        if(!pathFile.exists()) pathFile.createNewFile();
		
        FileUtils.writeStringToFile(pathFile,
                 new Gson().toJson(path) ,
                 Charset.defaultCharset());
                 
                 } catch (Exception e) {}
   
   
         }
	
	

	}

File: app/src/main/java/com/ardev/idroid/ui/main/MainActivity.kt
package com.ardev.idroid.ui.main


import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ObjectAnimator
import android.app.Activity
import androidx.activity.viewModels
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.text.TextUtils
import com.ardev.component.widget.breadcrumbs.model.IBreadcrumbItem
import com.ardev.idroid.ext.JsonUtils
import com.ardev.idroid.service.IndexServiceConnection
import com.ardev.idroid.service.IndexService
import com.ardev.idroid.ui.home.ProjectProvider
import androidx.annotation.NonNull
import com.ardev.builder.project.Project
import com.ardev.idroid.ext.ProjectUtils
import com.ardev.idroid.ui.home.ProjectManager
import com.google.android.material.transition.MaterialFade
import android.view.ViewGroup
import androidx.transition.TransitionManager
import com.ardev.idroid.ui.main.model.Editable
import com.ardev.idroid.ui.main.adapter.PageAdapter
import com.ardev.idroid.databinding.ActivityMainBinding
import com.ardev.idroid.ui.main.adapter.FileListAdapter
import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import com.google.gson.reflect.TypeToken
import com.ardev.idroid.ext.*
import com.ardev.idroid.app.MainViewModelProvider
import android.view.LayoutInflater
import android.view.View
import android.widget.Toast
import androidx.appcompat.widget.PopupMenu
import android.view.animation.AnticipateInterpolator
import androidx.core.graphics.Insets
import androidx.core.view.GravityCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.Observer
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.lifecycle.SavedStateViewModelFactory
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.drawerlayout.widget.DrawerLayout
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.ardev.component.widget.breadcrumbs.BreadcrumbsView
import com.google.android.material.tabs.TabLayoutMediator
import java.lang.reflect.Type
import com.ardev.idroid.base.BaseActivity
import com.google.gson.Gson
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets
import java.util.ArrayList
import java.util.List
import java.util.Arrays
import org.apache.commons.io.FileUtils

import java.util.concurrent.Executors
import com.google.android.material.appbar.MaterialToolbar
import java.io.File
import androidx.viewpager2.widget.CompositePageTransformer
import com.ardev.component.widget.breadcrumbs.DefaultBreadcrumbsCallback
import com.google.android.material.tabs.TabLayout
import com.ardev.component.widget.breadcrumbs.model.BreadcrumbItem
import com.ardev.idroid.R

class MainActivity: BaseActivity<ActivityMainBinding, MainViewModel>(), ProjectManager.OnProjectOpenListener {
	
	
private lateinit var mFilesAdapter: FileListAdapter 
private lateinit var mPagerAdapter: PageAdapter 
private lateinit var project: Project
private lateinit var mProjectManager: ProjectManager
private lateinit var mIndexServiceConnection: mIndexServiceConnection

var firstTimeInitTab = true
var firstTimeInitList = true

	
	override val viewmodel: MainViewModel by viewModels
    override fun getViewBinding() = ActivityMainBinding.inflate(layoutInflater)

	override fun onCreate(bundle: Bundle) {
		super.onCreate(bundle)
	MainViewModelProvider.init(viewmodel)
	
	 project = ProjectProvider.getProject()
	 mProjectManager = ProjectManager.getInstance()
     mProjectManager.addOnProjectOpenListener(this@MainActivity)
	  mIndexServiceConnection = IndexServiceConnection(viewmodel)
    
		initView()
		viewmodel.setFatherPath(project.getRootFile().path)
		
	if (!project.equals(mProjectManager.getCurrentProject())) openProject(project)
	}
	
	private fun initView() {
	with(binding) {
navView.addSystemWindowInsetToPadding(false, false, false, false)
pager2.addSystemWindowInsetToMargin(false, false, false, false)
pager2.addInsetTypeImeToPadding()

	setSupportActionBar(toolbar)
	supportActionBar.setDisplayHomeAsUpEnabled(true)
	
		val toggle = ActionBarDrawerToggle(this@MainActivity, drawer, toolbar, R.string.app_name, R.string.app_name)
		toggle.syncState()
		toggle.setDrawerSlideAnimationEnabled(true)
		drawer.addDrawerListener(toggle)
		
		breadcrumbsView.setCallback { item, position -> 
		viewmodel.setCurrentPath(getPath(position))
		 bkpListBreadcrumb(breadcrumbsView.getItems())
		
	}
}
		initRecyclerView()
		initTab()
		}
		
		private fun initRecyclerView(){
	with(binding) {
mFilesAdapter = FileListAdapter()
	with(list) {
 	adapter = mFilesAdapter
	setHasFixedSize(true)
	setItemAnimator(DefaultItemAnimator())
	}
mFilesAdapter.setOnClickListener { view, file -> 
			 if (file.isDirectory()) {
  	viewmodel.addBreadcrumbItem(file.path)
	viewmodel.setCurrentPath(file.path)
		} else {
	viewmodel.addEditor(Editable(file, false))
drawer.closeDrawer(GravityCompat.START)
}
}

mFilesAdapter.setOnLongClickListener { v, file -> 

if (file.isDirectory()) showFolderOptionsMenu(v, file)
else if((ProjectUtils.isLayoutXMLFile(file) || ProjectUtils.isValuesXMLFile(file)  || ProjectUtils.isDrawableXMLFile(file)) && file.isFile()) showFileOptionsMenu(v, file)

}
 }
 
 
 }
 
		override fun observeViewModel() {
 	
 	viewmodel.fatherPath.observe(this) { fatherPath -> 
       	supportActionBar.setTitle(File(fatherPath).name)
       		
      val breadcrumbsCache = File("${fatherPath}/app/build/.cache", ".breadcrumbs")
      
if (breadcrumbsCache.exists()) {
	try {
  val listBreadCrumb: ArrayList<BreadcrumbItem> =  JsonUtils.jsonToList(breadcrumbsCache.readText(), BreadcrumbItem::class.java)
   listBreadCrumb.forEach{binding.breadcrumbsView.addItem(it)}

	} catch (e: Exception) {}
		
		} else {
	
	binding.breadcrumbsView.addItem(BreadcrumbItem.createSimpleItem(File(fatherPath).name))
	
	}
            }
            
	viewmodel.currentPath.observe(this) { path -> 
	var _path: String = null
	if(firstTimeInitList) {
                firstTimeInitList = false
     val currentPathCache = File("${viewmodel.fatherPath.value}/app/build/.cache", ".currentpath")
if (currentPathCache.exists()) {
	try {
   _path =  Gson().fromJson(currentPathCache.readText(), String::class.java)
   
	} catch (e: Exception) {} 
	} else {
	_path = path
	}
	} else {
	_path = path
	}
	showList(_path)
	bkpCurrentPath(_path)
	}
	
	
	
	
	viewmodel.breadcrumbItem.observe(this) { file -> 
		val breadcrumbItem = BreadcrumbItem(viewmodel.listDir().value)
		breadcrumbItem.setSelectedItem(file.name)
  binding.breadcrumbsView.addItem(breadcrumbItem)
  bkpListBreadcrumb(binding.breadcrumbsView.getItems())
	}
		
	viewmodel.listEditor
                .observe(this) { files -> 
                
                    mPagerAdapter.submitList(files)
                    binding.tab.setVisibility( if(files.isEmpty()) View.GONE else View.VISIBLE)
                }
                
   viewmodel.currentTabPosition
                .observe(this) { position -> 
                if(firstTimeInitTab) {
                firstTimeInitTab = false
                 val positionFile = File("${viewmodel.fatherPath.value}/app/build/.cache", ".position")
if (positionFile.exists()) {
	try {
  val pos: Int = Gson().fromJson(positionFile.readText(), Int::class.java)
  binding.pager2.setCurrentItem(pos, false)
		} catch (e: Exception) {} 
		} else {
			 binding.pager2.setCurrentItem(position, false)
			}
              } else {
                
             binding.pager2.setCurrentItem(position, false)
				
               }
                 }
                 
    viewmodel.isIndexing
                .observe(this) { indexing -> 
                   
                }
                
        viewmodel.currentState
                .observe(this) { message -> 
                supportActionBar.setSubtitle(message) 
                
                }
		}
		
private fun getPath(depth: Int): String {
	if (depth == -1) depth = binding.breadcrumbsView.getItems().size() - 1
	val sbPath =  StringBuffer(viewmodel.fatherPath.value)
	for (i in 1..depth) {
		
		sbPath.append("/").append(v.breadcrumbsView.getItems().get(i).getSelectedItem())
	}
	return sbPath.toString()
}
 
 private fun showList(path: String ) {
   
    Executors.newSingleThreadExecutor().execute {
    
       val list = File(path).listFiles().toList()
         
       runOnUiThread{
     
	
                    
            if (list.isEmpty()) {
                binding.scrollingView.setVisibility(View.GONE)
                binding.emptyItem.setVisibility(View.VISIBLE)
            } else {
                binding.scrollingView.setVisibility(View.VISIBLE)
                binding.emptyItem.setVisibility(View.GONE)
                 mFilesAdapter.setList(list)
        viewmodel.setListCurrent(list)
        }
        
            
	
	}
	}
	}

private fun showFolderOptionsMenu(view: View, file: File) {
   
    val chars: Array<CharSequence> = arrayOf("Compartir", "Borrar")
	
	val dialog = MaterialAlertDialogBuilder(this, R.style.ThemeOverlay_MaterialAlertDialog_NoRounded)
		 
		dialog.setItems(chars) { d, item -> 

			 
		 }
		 dialog.show()
        
    }
private fun showFileOptionsMenu(view: View, file: File) {
   
     val chars: Array<CharSequence> = arrayOf("Vista Previa", "Editar")
    
		val dialog = MaterialAlertDialogBuilder(this, R.style.ThemeOverlay_MaterialAlertDialog_NoRounded)
		 
		dialog.setItems(chars) { d, item -> 
		when (item) {
		 0 -> {
		viewmodel.addEditor(Editable(file, true))
		binding.drawer.closeDrawer(GravityCompat.START)
			}
		1 -> {
		viewmodel.addEditor(Editable(file, false))
		binding.drawer.closeDrawer(GravityCompat.START)
			}
		}
			 
		 }
		 dialog.show()
        
    }
    
   private fun showTabMenu(tab: TabLayout.Tab) {
   val editable: Editable = viewmodel.listEditor.value.get(tab.getPosition())
  val isPreview = editable.isPreview()
   
   val menu =  PopupMenu(this, tab.view)
	with(menu.getMenu()) {
		add("Cerrar")
		add("Cerrar otros")
		add("Cerrar todos")
		}
		
		menu.setOnMenuItemClickListener{ item -> 
			 when (item.getTitle().toString()) {

	 "Cerrar" -> viewmodel.closeEditor(editable)
	"Cerrar otros" -> viewmodel.closeOthersEditor(editable)
	"Cerrar todos" -> viewmodel.closeAllEditor()
						
					 }
					 
					 true
				}
				menu.show()
   
   
   }
    
    
    
    private fun initTab() {
    
    
		
   mPagerAdapter = PageAdapter(supportFragmentManager, lifecycle)
        binding.pager2.setAdapter(mPagerAdapter)
        binding.pager2.setUserInputEnabled(false)
		binding.pager2.setPageTransformer { view, position -> 
		view.setAlpha(0.2f)
		view.setVisibility(View.VISIBLE)
		view.animate().alpha(1.0f).setDuration(view.getResources().getInteger(android.R.integer.config_shortAnimTime))
		}
        
        v.tab.addOnTabSelectedListener(object: TabLayout.OnTabSelectedListener() {
            override fun onTabUnselected(tab: TabLayout.Tab) {
                
            }

            override fun onTabReselected(tab: TabLayout.Tab) {
           showTabMenu(tab)
               
            }

            override fun onTabSelected(tab: TabLayout.Tab) {
               viewmodel.setCurrentTabPosition(tab.getPosition())

            }
        })
		
       TabLayoutMediator(binding.tab, binding.pager2, true, false) { tab, i -> 
			val isPreview = viewmodel.getListEditor().value.get(i).isPreview()

			val fileName = viewmodel.getListEditor().value.get(i).getFile().name
			
			tab.setText(fileName)
        	if(isPreview) tab.setIcon(R.drawable.ic_preview)
        	 
        	
        }.attach()
    
    
    
    }
    
    
     fun openProject(project: Project) {
        mIndexServiceConnection.setProject(project)
        viewmodel.setIndexing(true)
        val intent = Intent(this@MainActivity, IndexService::class.java)
        startService(intent)
        bindService(intent, mIndexServiceConnection, Context.BIND_IMPORTANT)
    }
    
     override fun onProjectOpen(project: Project) {
    
    
    }
    
    
   override fun onBackPressed() {
             if (binding.drawer.isDrawerOpen(GravityCompat.START)) {
            binding.drawer.closeDrawer(GravityCompat.START)
            } else {
            super.onBackPressed()
            }
	}
	
	

	  private fun bkpListBreadcrumb(bList: List<IBreadcrumbItem>) {
   try {
        val bciFile = File("${viewmodel.fatherPath.value}/app/build/.cache", ".breadcrumbs")
         val content  = GsonBuilder()
                .setPrettyPrinting()
                .create()
                .toJson( bList, TypeToken<List<BreadcrumbItem>>(){}.type)
        
        if(!bciFile.exists()) bciFile.createNewFile()
		if(!bList.isEmpty()){
       bciFile.writeText(content)
                 }
                 } catch (e: Exception) {}
   
   }
   
   private fun bkpCurrentPath(path: String) {
   try {
         val pathFile =  File("${viewmodel.fatherPath.value}/app/build/.cache", ".currentpath")
        
        if(!pathFile.exists()) pathFile.createNewFile()
		pathFile.writeText(Gson().toJson(path))
       
                 } catch (e: Exception) {}
   
   
         }
	
	

	}